x/avs/keeper/task.go

```diff
@@ -102,3 +110,264 @@ func (k Keeper) IterateTaskAVSInfo(ctx sdk.Context, fn func(index int64, taskInf
+	case types.TwoPhaseCommitOne:
+		if k.IsExistTaskResultInfo(ctx, info.OperatorAddress, info.TaskContractAddress, info.TaskId) {
+			return errorsmod.Wrap(
+				types.ErrResAlreadyExists,
+				fmt.Sprintf("SetTaskResultInfo: task result is already exists, "+
+					"OperatorAddress: %s (TaskContractAddress: %s),(Task ID: %d)",
+					info.OperatorAddress, info.TaskContractAddress, info.TaskId),
+			)
+		}
+		if info.TaskResponseHash != "" || info.TaskResponse != nil {
+			return errorsmod.Wrap(
+				types.ErrParamNotEmptyError,
+				fmt.Sprintf("SetTaskResultInfo: invalid param TaskResponseHash: %s (TaskResponse: %d)",
+					info.TaskResponseHash, info.TaskResponse),
+			)
+		}
+		// check epoch，The first stage submission must be within the response window period
+		// #nosec G115
+		if epoch.CurrentEpoch > int64(task.StartingEpoch)+int64(task.TaskResponsePeriod) {
+			return errorsmod.Wrap(
+				types.ErrSubmitTooLateError,
+				fmt.Sprintf("SetTaskResultInfo:submit  too late, CurrentEpoch:%d", epoch.CurrentEpoch),
+			)
+		}
+		infoKey := assetstype.GetJoinedStoreKey(info.OperatorAddress, info.TaskContractAddress,
+			strconv.FormatUint(info.TaskId, 10))
+		store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixTaskResult)
+		bz := k.cdc.MustMarshal(info)
+		store.Set(infoKey, bz)
+		return nil
+
+	case types.TwoPhaseCommitTwo:
+		// check task response
+		if info.TaskResponseHash == "" || info.TaskResponse == nil {
+			return errorsmod.Wrap(
+				types.ErrNotNull,
+				fmt.Sprintf("SetTaskResultInfo: invalid param TaskResponseHash: %s (TaskResponse: %d)",
+					info.TaskResponseHash, info.TaskResponse),
+			)
+		}
+		// check parameters
+		res, err := k.GetTaskResultInfo(ctx, info.OperatorAddress, info.TaskContractAddress, info.TaskId)
+		if err != nil || res.OperatorAddress != info.OperatorAddress ||
+			res.TaskContractAddress != info.TaskContractAddress ||
+			res.TaskId != info.TaskId || !bytes.Equal(res.BlsSignature, info.BlsSignature) {
+			return errorsmod.Wrap(
+				types.ErrInconsistentParams,
+				fmt.Sprintf("SetTaskResultInfo: invalid param OperatorAddress: %s ,(TaskContractAddress: %s)"+
+					",(TaskId: %d),(BlsSignature: %s)",
+					info.OperatorAddress, info.TaskContractAddress, info.TaskId, info.BlsSignature),
+			)
+		}
+		//  check epoch，The second stage submission must be within the statistical window period
+		// #nosec G115
+		if epoch.CurrentEpoch <= int64(task.StartingEpoch)+int64(task.TaskResponsePeriod) ||
+			epoch.CurrentEpoch > int64(task.StartingEpoch)+int64(task.TaskResponsePeriod)+int64(task.TaskStatisticalPeriod) {
+			return errorsmod.Wrap(
+				types.ErrSubmitTooLateError,
+				fmt.Sprintf("SetTaskResultInfo:submit  too late, CurrentEpoch:%d", epoch.CurrentEpoch),
+			)
+		}
+
+		// check hash
+		taskResponseDigest := crypto.Keccak256Hash(info.TaskResponse)
+		hashWithoutPrefix := strings.TrimPrefix(taskResponseDigest.String(), "0x")
+		if hashWithoutPrefix != info.TaskResponseHash {
```

Comment:
it seems we don't need to ask operator to provide responsehash since we could get the hash from response

Reply:
yes

Reply:
This parameter hash is optional